src/Func.cc

```diff
@@ -454,6 +474,49 @@ void BroFunc::AddBody(Stmt* new_body, id_list* new_inits, int new_frame_size,
 	sort(bodies.begin(), bodies.end());
 	}
 
+void BroFunc::AddClosure(std::shared_ptr<id_list> ids, Frame* f)
+	{
+	// Order matters here.
+	this->SetOuterIDs(ids);
+	this->SetClosureFrame(f);
+	}
+
+void BroFunc::SetClosureFrame(Frame* f)
+	{
+	if (this->closure)
+		reporter->InternalError
+			("Tried to override closure for BroFunc %s.", this->Name());
+
+	this->closure = f ? f->SelectiveClone(this->outer_ids.get()) : nullptr;
+	}
+
+Val* BroFunc::DoClone()
+	{
+	// A BroFunc could hold a closure. In this case a clone of it must copy this
+	// store a copy of this closure.
+	if ( ! this->closure )
+		{
+		return Func::DoClone();
+		}
+	else
+		{
+		BroFunc* other = new BroFunc();
+
+		other->bodies = this->bodies;
+		other->scope = this->scope;
+		other->kind = this->kind;
+		other->type = this->type;
```

Comment:
It isn't so nice to clone the member variables of the parent class here. If the parent class ever changes, it's likely that one would forget to update this code. I realize that calling `Func::DoClone()` won't work, as it optimizes for the cases that it doesn't need to maintain state. But how about at least adding a protected method `CopyState()` or so to `Func` that takes care of the copying its state over into a new instance. That way the logic is associated with that class at least.

Reply:
2nd note is here is that `Func::~Func` unrefs `type`, so it needs to be ref'ed on clone.